ntoskrnl/se/debug.c

```diff
@@ -0,0 +1,312 @@
+        DbgPrint("Ace->AccessMask -> 0x%08lx\n", Ace->AccessMask);
+
+        Sid = SepGetSidFromAce(Ace->Header.AceType, Ace);
+        ASSERT(Sid);
+        RtlConvertSidToUnicodeString(&SidString, Sid, TRUE);
+        DbgPrint("Ace SID -> %wZ\n", &SidString);
+        RtlFreeUnicodeString(&SidString);
+
+        DbgPrint("Ace->Header.AceSize -> %u\n", Ace->Header.AceSize);
+        DbgPrint("Ace->Header.AceFlags:\n");
+        SepDumpAceFlags(Ace->Header.AceFlags);
+    }
+}
+
+/**
+ * @brief
+ * Dumps debug info of an Access Control List (ACL).
+ */
+static
+VOID
+SepDumpAclInfo(
+    _In_ PACL Acl,
+    _In_ BOOLEAN IsSacl)
+{
+    /* Dump relevant info */
+    DbgPrint("================== %s DUMP INFO ==================\n", IsSacl ? "SACL" : "DACL");
+    DbgPrint("Acl->AclRevision -> %u\n", Acl->AclRevision);
+    DbgPrint("Acl->AclSize -> %u\n", Acl->AclSize);
+    DbgPrint("Acl->AceCount -> %u\n", Acl->AceCount);
+
+    /* Dump all the ACEs present on this ACL */
+    SepDumpAces(Acl);
+}
+
+/**
+ * @brief
+ * Dumps control flags of a security descriptor to the debugger.
+ */
+static
+VOID
+SepDumpSdControlInfo(
+    _In_ SECURITY_DESCRIPTOR_CONTROL SdControl)
+{
+#define SD_CONTROL_PRINT(x) \
+    if (SdControl & x)      \
+    {                       \
+        DbgPrint(#x "\n");  \
+    }
+
+    SD_CONTROL_PRINT(SE_OWNER_DEFAULTED);
+    SD_CONTROL_PRINT(SE_GROUP_DEFAULTED);
+    SD_CONTROL_PRINT(SE_DACL_PRESENT);
+    SD_CONTROL_PRINT(SE_DACL_DEFAULTED);
+    SD_CONTROL_PRINT(SE_SACL_PRESENT);
+    SD_CONTROL_PRINT(SE_SACL_DEFAULTED);
+    SD_CONTROL_PRINT(SE_DACL_UNTRUSTED);
+    SD_CONTROL_PRINT(SE_SERVER_SECURITY);
+    SD_CONTROL_PRINT(SE_DACL_AUTO_INHERIT_REQ);
+    SD_CONTROL_PRINT(SE_SACL_AUTO_INHERIT_REQ);
+    SD_CONTROL_PRINT(SE_DACL_AUTO_INHERITED);
+    SD_CONTROL_PRINT(SE_SACL_AUTO_INHERITED);
+    SD_CONTROL_PRINT(SE_DACL_PROTECTED);
+    SD_CONTROL_PRINT(SE_SACL_PROTECTED);
+    SD_CONTROL_PRINT(SE_RM_CONTROL_VALID);
+    SD_CONTROL_PRINT(SE_SELF_RELATIVE);
+#undef SD_CONTROL_PRINT
+}
+
+/**
+ * @brief
+ * Dumps each security identifier (SID) of an access token to debugger.
+ */
+static
+VOID
+SepDumpSidsOfToken(
+    _In_ PSID_AND_ATTRIBUTES Sids,
+    _In_ ULONG SidCount)
+{
+    ULONG SidIndex;
+    UNICODE_STRING SidString;
+
+    /* Loop all SIDs and dump them */
+    for (SidIndex = 0; SidIndex < SidCount; SidIndex++)
+    {
+        RtlConvertSidToUnicodeString(&SidString, Sids[SidIndex].Sid, TRUE);
+        DbgPrint("%lu# %wZ\n", SidIndex, &SidString);
+        RtlFreeUnicodeString(&SidString);
+    }
+}
+
+/* PUBLIC FUNCTIONS ***********************************************************/
+
+/**
+ * @brief
+ * Dumps debug information of a security descriptor to the debugger.
+ */
+VOID
+SepDumpSdDebugInfo(
+    _In_ PISECURITY_DESCRIPTOR SecurityDescriptor)
+{
+    UNICODE_STRING SidString;
+    PSID OwnerSid, GroupSid;
+    PACL Dacl, Sacl;
+
```

Comment:
Please guard this function (and the ones below) against invalid input.
At least do a NULL check, and if it's possible (not sure?) try to do further checks against garbage input.